neet/sensitivity.py

```diff
@@ -35,33 +39,178 @@ def sensitivity(net, state):
+                          when states argument is not None.
+    """
+    N = net.size
+    Q = np.zeros((N,N))
+
+    if (states is not None) or (weights is not None) or (not hasattr(net,'table')):
+        # explicitly calculate difference matrix for each state
+    
+        # optionally pre-calculate transitions
+        if calc_trans:
+            trans = list(transitions(net))
+        else:
+            trans = None
+    
+        # currently changes state generators to lists.
+        # is there a way to avoid this?
+        if states is None:
+            states = list(net.state_space())
+        else:
+            states = list(states)
+
+        if weights is None:
+            weights = np.ones(len(states))
+        else:
+            weights = list(weights)
+
+        if np.shape(weights) != (len(states),):
+            raise(ValueError("Weights must be a 1D array with length same as states"))
+
+        norm = np.sum(weights)
+        for i,state in enumerate(states):
+            Q += weights[i] * difference_matrix(net, state, trans) / norm
+
+    else: # make use of sparse connectivity to be more efficient
+        state0 = np.zeros(N,dtype=int)
+
+        for i in range(N):
+            nodesInfluencingI = _connections(net,i)
+            for jindex,j in enumerate(nodesInfluencingI):
+            
+                # for each state of other nodes, does j matter?
+                otherNodes = list(copy.copy(nodesInfluencingI))
+                otherNodes.pop(jindex)
+                otherNodeStates = _states_limited(otherNodes,state0)
+                for state in otherNodeStates:
+                    # (might be able to do faster by calculating transitions once
+                    #  for each i)
+                    # (also we only need the update for node i)
+                    # start with two states, one with j on and one with j off
+                    jOff = copy.copy(state)
+                    jOff[j] = 0
+                    jOffNext = net.update(jOff)[i]
+                    jOn = copy.copy(state)
+                    jOn[j] = 1
+                    jOnNext = net.update(jOn)[i]
+                    # are the results different?
+                    Q[i,j] += (jOffNext + jOnNext)%2
+                Q[i,j] /= float(len(otherNodeStates))
+            
+    return Q
+
+def lambdaQ(net,**kwargs):
+    """
+    Calculate sensitivity eigenvalue, the largest eigenvalue of the sensitivity
+    matrix average_difference_matrix.
+    
+    This is analogous to the eigenvalue calculated in
+    
+        Pomerance, A, E Ott, M Girvan, and W Losert. ``The Effect of Network
+        Topology on the Stability of Discrete State Models of Genetic Control.''
+        Proc. Natl. Acad. Sci. USA 106, no. 20 (2009): 8209-14.
+        http://www.pnas.org/content/106/20/8209.full.
+    """
+    Q = average_difference_matrix(net,**kwargs)
+    return max(np.sort(abs(linalg.eigvals(Q))))
```

Comment:
Two questions here:
1. Why take the absolute value? It's possible you could get a negative eigenvalue (which you force to positive). It seems to me that that would be undesirable.
2. Why sort the array? `max` will pull out the largest value either way.

Reply:
1. Perron-Frobenius says the eigenvalue we want (with largest magnitude) will be real and positive.  I used abs because I wasn't sure how max would work with complex numbers and I was worried there might be cases where numerical issues could lead to small imaginary parts even when the one we want is supposed to be real.
2. Ha, oops.  Good point :)